src/Paprika/Merkle/MerkleNode.cs

```diff
@@ -0,0 +1,271 @@
+using System.Buffers.Binary;
+using System.Runtime.InteropServices;
+using Paprika.Crypto;
+using Paprika.Data;
+
+namespace Paprika.Merkle;
+
+public static class Node
+{
+    public enum Type : byte
+    {
+        Leaf,
+        Extension,
+        Branch,
+    }
+
+    public static ReadOnlySpan<byte> ReadFrom(ReadOnlySpan<byte> source, out Type nodeType, out Leaf leaf, out Extension extension, out Branch branch)
+    {
+        leaf = new Leaf();
+        extension = new Extension();
+        branch = new Branch();
+
+        // TODO: It would be nice to not read the header twice:
```

Comment:
Thank you for this comment. I think we can amend it in the future, but as this does not impact the data format, we should be good.